src/storage/src/types/errors.proto

```diff
@@ -34,10 +40,32 @@ message ProtoSourceError {
     ProtoSourceErrorDetails error = 2;
 }
 
+message ProtoUpsertValueError {
+    ProtoDataflowError inner = 1;
+    mz_repr.row.ProtoRow for_key = 2;
+}
+
+message ProtoUpsertError {
+    oneof kind {
+	ProtoDecodeError key_decode = 1;
+	ProtoUpsertValueError value = 2;
+    }
+}
+
+message ProtoEnvelopeError {
+    oneof kind {
+	string debezium = 1;
+	ProtoUpsertError upsert = 2;
+	string flat = 3;
+    }
+}
+
+
 message ProtoDataflowError {
     oneof kind {
         ProtoDecodeError decode_error = 1;
         mz_expr.scalar.ProtoEvalError eval_error = 2;
         ProtoSourceError source_error = 3;
+	ProtoEnvelopeError envelope_error = 4;
```

Comment:
because we are going to need to change `DataflowError` again to clean this all up, as mentioned here: https://github.com/MaterializeInc/materialize/pull/14209#issuecomment-1215054123, and we may well use `envelope_error` again, it may be nice to change this to:
```
ProtoEnvelopeErrorV1 envelope_error_v1
```

so we can reserve `ProtoEnvelopeError` for later. Both structs will likely have to co-exist, because whatever change we make to the storage error structures will have to be backwards compatible going forward